source/common/stats/lazy_init.h

```diff
@@ -0,0 +1,73 @@
+#pragma once
+
+#include "source/common/common/thread.h"
+
+namespace Envoy {
+namespace Stats {
+
+template <typename StatsStructType> class LazyableInterface {
+public:
+  // Helper operators to get-or-create and return the StatsStructType object.
+  virtual StatsStructType* operator->() PURE;
+  virtual StatsStructType& operator*() PURE;
+  virtual ~LazyableInterface() {}
+};
+
+/**
+ * Lazy-initialization wrapper for StatsStructType, intended for deferred instantiation of a block
+ * of stats that might not be needed in a given Envoy process.
+ *
+ * This class is thread-safe -- instantiations can occur on multiple concurrent threads.
+ */
+template <typename StatsStructType> class LazyInit : public LazyableInterface<StatsStructType> {
+public:
+  // Capture the stat names object and the scope with a ctor, that can be used to instantiate a
+  // StatsStructType object later.
+  // Caller should make sure scope and stat_names outlive this object.
+  LazyInit(const typename StatsStructType::StatNameType& stat_names, Stats::Scope& scope)
+      : inited_(Stats::Utility::gaugeFromElements(
+            scope, {Stats::DynamicName(StatsStructType::typeName()), Stats::DynamicName("inited")},
+            Stats::Gauge::ImportMode::NeverImport)),
```

Comment:
we might want to import this, right? Otherwise I think we will lose all stat values on a hot-restart.